src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java

```diff
@@ -117,4 +212,111 @@ public boolean checkPipe() {
+  }
+
+  public void setSetpointX(double point) {
+    PIDVision.setSetpoint(point);
+  }
+
+  public void setSetpointY(double point) {
+    PIDVisionY.setSetpoint(point);
+  }
+
+  public void setSetpointTheta(double point) {
+    PIDVisionTheta.setSetpoint(point);
+  }
+
+// megatag tx and ty, units in meters
+  public Pose2d getPose2D() {
+    return new Pose2d(Math.abs(getBotPose()[0]), Math.abs(getBotPose()[1]), Rotation2d.fromDegrees(getBotPose()[5]));
+  }
+
+  public double[] getBlueBotPose(){
+    double[] blueBotPose = this.blueBotPose.getDoubleArray(new double[11]);
+    double[] colorBotPose = new double[12];
+
+    colorBotPose[0] = blueBotPose[0];
+    colorBotPose[1] = blueBotPose[1];
+    colorBotPose[2] = blueBotPose[2];
+    colorBotPose[3] = blueBotPose[3];
+    colorBotPose[4] = blueBotPose[4];
+    colorBotPose[5] = blueBotPose[5];
+    colorBotPose[6] = blueBotPose[6];
+    colorBotPose[7] = blueBotPose[7];
+    colorBotPose[8] = blueBotPose[8];
+    colorBotPose[10] = blueBotPose[10];
+
+    if(hasTarget()){
+      colorBotPose[9] = blueBotPose[9];
+      colorBotPose[11] = getTID()+1;
+
+    }
+    
+
+    return colorBotPose; 
+  }
+
+  public double[] getTargetSpace(){
+    double[] targetSpace = this.targetSpace.getDoubleArray(new double[6]);
+    return targetSpace;
+  } 
+
+  public int getTID(){
+  
+    int tid = (int)this.tid.getDouble(-1)-1;
+    
+    return tid;
+    
+  }
+
+  public void setData(){
+    double[] temp = getBlueBotPose();
+    Pose2d currentPose = VisionContainer.drivebase.getPose();
+    double distance = Math.sqrt(Math.pow(temp[0]- currentPose.getX(),2) + Math.pow(temp[1]- currentPose.getY() ,2)); //Calculates distance from Apriltag to robot
+    double distancefromAprilTag = 0.8; //0.8
+    double distancefromLimeLight = 2.5; //2.5
+
+    if(hasTarget()){
+       
+        double xyStds= 0;
+        double degStds = 0;
+        Matrix<N3, N1> stds = new Matrix<N3, N1>(Nat.N3(), Nat.N1());
+        LimelightHelpers.PoseEstimate limelightMeasurement = LimelightHelpers.getBotPoseEstimate_wpiBlue("limelight");
+        if(limelightMeasurement.tagCount >= 2 ){ //Checks if Limelight sees 2 Apriltag
+            xyStds = 0.5; 
+            degStds = 6;
+        
+        }
+        else if ((temp[9] < distancefromLimeLight) && (distance < distancefromAprilTag)) { //Checks if within distance of apriltag and limelight
+            xyStds = 1.0;
+            degStds = 12;
+
+        }
+        else{
+          return;
+        }
+          stds.set(0,0,xyStds);
+          stds.set(1,0,xyStds);
+          stds.set(2,0, degStds);
+          Rotation2d degree = new Rotation2d(temp[5]* Math.PI / 180);
+          Pose2d newPose = new Pose2d(temp[0],temp[1],degree); //creates new pose2d with limelight data
+          
+          VisionContainer.drivebase.addActualVisionReading(newPose ,Timer.getFPGATimestamp() - (temp[6]/1000.0),stds); //Changes standard dev base on apriltags and drive
```

Comment:
We should comment this line of code first. It will crash the main branch when VisionContainer's drivebase is not initialized.

Should use the GameRobotContainer's drivebase instead.

Reply:
Or pass in as a parameter to this subsystem.